puppi/firmware/simple_puppi.cpp

```diff
@@ -0,0 +1,79 @@
+#include "simple_puppi.h"
+#include <cassert>
+#ifndef __SYNTHESIS__
+#include <cstdio>
+#endif
+
+typedef ap_uint<7> tk2em_dr_t;
+typedef ap_uint<10> tk2calo_dr_t;
+typedef ap_uint<10> em2calo_dr_t;
+typedef ap_uint<12> tk2calo_dq_t;
+typedef ap_uint<12> mu2trk_dr_t;
+
+weight_t puppiweight(int iWeight){
+  static weight_t _table[PUPPI_TABLE_SIZE];
+  lut_puppiweight_init<weight_t,PUPPI_TABLE_SIZE>(_table);
+  return _table[iWeight];
+}
+
+int dr2_int(etaphi_t eta1, etaphi_t phi1, etaphi_t eta2, etaphi_t phi2) {
+    etaphi_t deta = (eta1-eta2);
+    etaphi_t dphi = (phi1-phi2);
+    return deta*deta + dphi*dphi;
+}
+
+void simple_puppi_hw(PFChargedObj pfch[NTRACK], PFNeutralObj pfallne[NNEUTRALS], z0_t Z0) {
+
+    const z0_t DZMAX = 256;
+    const int DR2MAX = 8404; // 0.4 cone
+
+    #pragma HLS ARRAY_PARTITION variable=pfch complete
+    #pragma HLS ARRAY_PARTITION variable=pfallne complete
+    #pragma HLS pipeline II=1
+
+    ap_uint<32> eToAlphas[NNEUTRALS];
+    ap_uint<8> weights[NNEUTRALS];
+    #pragma HLS ARRAY_PARTITION variable=eToAlphas complete
+    #pragma HLS ARRAY_PARTITION variable=weights complete    
+
+    for (int in = 0; in < NNEUTRALS; ++in){ 
+      eToAlphas[in] = 0; weights[in] = 0; 
+    }
+
+    for (int in = 0; in < NNEUTRALS; ++in) {
+        
+        int sum = 0;
+        for (int it = 0; it < NTRACK; ++it) {
+            // std::cout << "pfch[it].hwPt = " << pfch[it].hwPt << std::endl;
+
+            if ((Z0 - pfch[it].hwZ0 > DZMAX) || (Z0 - pfch[it].hwZ0 < -DZMAX)) continue; // if track is PV
+            int dr2 = dr2_int(pfch[it].hwEta, pfch[it].hwPhi, pfallne[in].hwEta, pfallne[in].hwPhi); // if dr is inside puppi cone
+            if (dr2 <= DR2MAX) {
+                ap_uint<9> dr2short = dr2 >> 5; // why?
+                int pt2 = pfch[it].hwPt*pfch[it].hwPt;
+                int term = (std::min(pt2 >> 5, 131071) << 15)/(dr2short > 0 ? dr2short : ap_uint<9>(1));
+                sum += term;
+            }
+        }    
+        eToAlphas[in] = sum >> 10;
```

Comment:
in line 55 you do the sums with 10 more extra bits, and then trash away the bits at the end in line 58
given that the number of items is small, you could already remove some bits before the addition.
I tired e.g. doing `>>7` of term before the addition, and `>>3` of the sum afterwards, and it saves ~10% of the LUTs and Registers in the algo, still giving the same answer. This may be a bit aggressive if there are more than 2^3=8 tracks from the PV, but probably `>>5` is ok

Reply:
ok, I went to `>>5` -- it saved about 3% of FFs/LUTs, seems like a good compromise